src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapedCraftingRecipeBuilder.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.item.recipe.crafting;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableTable;
+import com.google.common.collect.Lists;
+import it.unimi.dsi.fastutil.chars.Char2ObjectArrayMap;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.recipe.crafting.ShapedCraftingRecipe;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+public final class SpongeShapedCraftingRecipeBuilder implements ShapedCraftingRecipe.Builder {
+
+    private final List<String> aisle = Lists.newArrayList();
+    private final Map<Character, Predicate<ItemStackSnapshot>> ingredientMap = new Char2ObjectArrayMap<>();
+    private ItemStackSnapshot result = ItemStackSnapshot.NONE;
+
+    @Override
+    public ShapedCraftingRecipe.Builder aisle(String... aisle) {
+        this.aisle.clear();
+        ingredientMap.clear();
+
+        if (aisle != null) {
+            Collections.addAll(this.aisle, aisle);
+        }
+
+        return this;
+    }
+
+    @Override
+    public ShapedCraftingRecipe.Builder where(char symbol, Predicate<ItemStackSnapshot> ingredient) throws IllegalArgumentException {
+        checkState(!this.aisle.isEmpty(), "aisle must be set before setting aisle symbols");
+
```

Comment:
Thanks @ziceptor, I must've missed that. The question stays, though: Which empty lines are distracting? Which should I keep/remove?